### PR TITLE
Create modular versions of extensions

### PR DESCRIPTION
**Pull request by jodeleeuw**
Converts extensions to ES6 classes and modules.

### CODE REVIEW

File: packages/extension-webgazer/src/index.ts
Comment (bjoluc) [inline packages/extension-webgazer/src/index.ts:48]: You can destructure params in place and at the same time set default values:
```suggestion
  initialize({ round_predictions = true, auto_initialize = false, sampling_interval = 34 }: InitializeParameters): Promise<void> {
```

If you'd like to keep the params object:
```suggestion
  initialize(params: InitializeParameters): Promise<void> {
    // setting default values for params if not defined
    params = {
      round_predictions: true,
      auto_initialize: false,
      sampling_interval: 34,
      ...params
    };
```
https://www.freecodecamp.org/news/es5-to-esnext-heres-every-feature-added-to-javascript-since-2015-d0c255e13c6e/#default-parameters

File: packages/extension-webgazer/src/index.ts
Comment (bjoluc) [inline packages/extension-webgazer/src/index.ts:40]: If you already know the type of something here, add it with `:` – otherwise, TypeScript uses `any`.

File: packages/extension-webgazer/src/index.ts
Comment (bjoluc) [inline packages/extension-webgazer/src/index.ts:72]: Move this to the top of the method (readability)?

File: packages/jspsych/src/JsPsych.ts
Comment (bjoluc) [inline packages/jspsych/src/JsPsych.ts:445]: You removed the implicit return of the arrow function and now it returns undefined. `Promise.all()` needs a list of promises, but gets a list of `undefined`s here :D
```suggestion
        extensions.map((extension) => 
          this.extensions[extension.type.info.name].initialize(extension.params || {})
        )
```

File: packages/jspsych/src/JsPsych.ts
Comment (bjoluc) [inline packages/jspsych/src/JsPsych.ts:540]: I'd love to see a `for (const extension of trial.extensions)` here though ;)

File: packages/jspsych/src/JsPsych.ts
Comment (bjoluc) [inline packages/jspsych/src/JsPsych.ts:571]: And here too, obviously...

File: packages/jspsych/src/modules/extensions.ts
Comment (bjoluc) [inline packages/jspsych/src/modules/extensions.ts:14]: Great! :tada: 
Instead of using comments above these methods in the webgazer extension, it would make sense to add JsDoc comments here so we get comments for all future extensions as well and help extension authors doing their jobs :)

File: packages/jspsych/tests/extensions/extensions.test.ts
Comment (bjoluc) [inline packages/jspsych/tests/extensions/extensions.test.ts:14]:
```suggestion
    const jsPsych = initJsPsych({
```

File: packages/jspsych/tests/extensions/extensions.test.ts
Comment (bjoluc) [inline packages/jspsych/tests/extensions/extensions.test.ts:17]: How does this approach differ from awaiting `jsPsych.run()` and then asserting initFunc to have been called?

**Review by bjoluc — CHANGES_REQUESTED** [review]
Great work so far!

File: packages/extension-webgazer/src/index.ts
Comment (jodeleeuw) [inline packages/extension-webgazer/src/index.ts:36]: Can I provide jsDoc hints for the parameters that are nested in the InitializeParameters object?

File: packages/jspsych/src/JsPsych.ts
Comment (jodeleeuw) [reply packages/jspsych/src/JsPsych.ts]: Thanks! I added the `{}` because I was adding some other statements in this during testing. I didn't even realize that there was an implicit return!

File: packages/extension-webgazer/src/index.ts
Comment (jodeleeuw) [inline packages/extension-webgazer/src/index.ts:44]: Can I provide jsdoc hints for these parameters that are nested in the params object?

Comment (jodeleeuw) [reply packages/extension-webgazer/src/index.ts:44]: Oooh, maybe I do that by applying them to the Type declaration?

Comment (bjoluc) [reply packages/extension-webgazer/src/index.ts:64]: Exactly!

File: packages/extension-webgazer/src/index.ts
Comment (bjoluc) [inline packages/extension-webgazer/src/index.ts:55]: Great! In these cases, TS infers the boolean type automatically (type widening):
```suggestion
  private initialized = false;
  private currentTrialStart:number;
  private activeTrial = false;
```

File: packages/extension-webgazer/src/index.ts
Comment (bjoluc) [inline packages/extension-webgazer/src/index.ts:70]: These does not look necessary – aren't they `false` already?
```suggestion
```

**Review by bjoluc — APPROVED** [review]

**Comment by bjoluc** [comment]
I think I told Becky, but I didn't tell you (sorry): I started using prettier when almost all files had to be moved and modified anyway, so you could use https://github.com/prettier/prettier-vscode to auto-format your code now. Maintainers typically set up a pre-commit hook that runs prettier automatically (e.g. using Husky), but I didn't do that yet; If you agree, I can push that to modularization once this is merged.

**Comment by jodeleeuw** [comment]
@bjoluc @becky-gilbert I feel like this is in pretty good shape now. The main issue remaining is that I haven't actually tested the webgazer extension, since (a) I'm not sure if there is a smooth way to use webgazer in jest tests and (b) writing mocks for webgazer sounds painful.

What I did during development was manually test things with the plugins I had created. If I want to do that then either I should wait for @becky-gilbert to update those plugins or I should do that update myself.

**Comment by becky-gilbert** [comment]
Hi @jodeleeuw, I've been going through the plugins and example files in alphabetical order, but I can convert the webgazer plugins now and commit those separately so that you can test this. Do you want me to just push the plugins into the modularization branch or do a pull request?

**Comment by jodeleeuw** [comment]
@becky-gilbert I think you could just push them directly into the `modularization-extensions` branch. Even if you just do one of them and push it then I can see the changes and match them for the rest. I was cautious about doing it because I haven't been working on that piece of it at all yet.

**Comment by becky-gilbert** [comment]
@jodeleeuw ok will do ASAP!

**Comment by becky-gilbert** [comment]
@jodeleeuw I just pushed a commit to `modularization-extensions` with the modularized webgazer plugins. I also modified the `examples/webgazer.html` file so that you can use it for testing, and included the modularized `html-button-response` plugin (the example file uses both `html-button-response` and `html-keyboard-response`, and the latter has already been modularized).

The webgazer plugins are compiling successfully, but I couldn't test whether there are runtime errors. I tried running the webgazer.html example file but got stuck because I couldn't figure out how to initialize jsPsych with the webgazer extension, or how to add the extension to a trial. I assume this now needs to be done with a variable rather than the 'webgazer' string?

Also, I commented out the webgazer bits in the example file so that you can see that it runs without the extension. So I think what you need to do is: (1) fix the initialization with webgazer, (2) fix the extensions parameter for the trial object, and (3) add the webgazer plugin trials back onto the timeline.

**Comment by jodeleeuw** [comment]
Thanks @becky-gilbert! I'm having trouble running the `build` command. I'm running `npm run build` from the root of the repository. Is that what I should be doing, or should I do something else?

**Comment by becky-gilbert** [comment]
@jodeleeuw I've been running `npm run build` from the root of each package directory needed for the example file, so in: 
* `packages/jspsych`
* `packages/extension-webgazer`
* `packages/plugin-html-button-response`

etc.
Does that work for you?

**Comment by jodeleeuw** [comment]
Yep! I forgot that most of the plugins won't build correctly yet. This makes sense!

**Comment by jodeleeuw** [comment]
I am running into an issue here that I am hoping @bjoluc knows how to fix...

`webgazer.js` has a method to invoke a callback when a new gaze prediction is made. My extension hooks into this:

https://github.com/jspsych/jsPsych/blob/d2e92ec1a6c9ad420754cf9f56aa96e50ffd9a3b/packages/extension-webgazer/src/index.ts#L200

I pass this function as the callback, defined in the extension class as a `private` function:

https://github.com/jspsych/jsPsych/blob/d2e92ec1a6c9ad420754cf9f56aa96e50ffd9a3b/packages/extension-webgazer/src/index.ts#L291-L310

But it looks like the scope of `this` no longer refers to the instance of the extension when this function is invoked. Instead `this` is `undefined`. I'm not sure how to solve this. Any ideas?

File: packages/extension-webgazer/rollup.config.mjs
Comment (bjoluc) [inline packages/extension-webgazer/rollup.config.mjs]: I liked `jsPsychExtensionWebgazer` very much for consistency in the global scope :smiling_face_with_tear: I think it's nice to see whether something is an extension already before you read to the end of it; I'd personally rate consistency of global var names higher than consistency between global vars and class names. What do you think?

Comment (jodeleeuw) [reply packages/extension-webgazer/rollup.config.mjs]: That's fine with me. I switched it because the plugins don't have a corresponding identifier, but since plugins are much more common I'm OK with having the identifier only on the extensions.

**Comment by jodeleeuw** [comment]
> @jodeleeuw ES6 classes are just syntactic sugar for the old-style function-based class concept and do not behave much differently. Member functions have the same old `this` behavior as any other `function`. When you detach the member function from the class instance and call it on its own, `this` will be `undefined`. Two options here:
> 
> * Use an arrow function for the callback:
>   ```ts
>   private handleGazeDataUpdate = (gazeData, elapsedTime) => { 
>   ```
> * Bind this to the function – I like to use [auto-bind](https://github.com/sindresorhus/auto-bind) to do it for all function members and be safe (whatever users may do with the functions, `this` will be the class instance then). `auto-bind` is already a dependency of the `jspsych` package, so you can just
>   ```ts
>   import autoBind from "auto-bind";
>   ```

Got it! For some reason I thought that the way I was declaring it was equivalent to an arrow function, but I had no good reason for thinking that. I've resolved it now.

**Comment by bjoluc** [comment]
@jodeleeuw Would you like to squash-merge this into `modularization`? We could also pull the commit into #2078 then.

File: packages/jspsych/src/modules/extensions.ts
Comment (bjoluc) [inline packages/jspsych/src/modules/extensions.ts:3]: There wouldn't be a need to export this then, am I right?
In this case, we could also do
```suggestion
type JsPsychExtensionParameters = Record<string, any>;
```

**Review by bjoluc — COMMENTED** [review]
Some more comments after going over this again...

File: packages/extension-webgazer/src/index.ts
Comment (jodeleeuw) [reply packages/extension-webgazer/src/index.ts:11]: My thought was that `JsPsychExtensionParameters` might, in the future, have some shared property across extensions so it would be good to model this now. But admittedly I don't have any idea about what those parameters would be, and I thought of this prior to adding the extension-specific types in the webgazer extension. 

So I think it does make sense to just drop this entirely.

File: packages/jspsych/src/modules/extensions.ts
Comment (bjoluc) [inline packages/jspsych/src/modules/extensions.ts:6]: That demands a non-static info member. Intended?

Comment (jodeleeuw) [reply packages/jspsych/src/modules/extensions.ts:6]: Oops. Not intended. I was having some git configuration issues and didn't realize this snuck into the commit.

**Comment by jodeleeuw** [comment]
@bjoluc I think this is ready to merge into `modularization`

**Pull request merged**
